sdks/python/apache_beam/runners/common.py

```diff
@@ -183,6 +184,8 @@ def __init__(self, obj_to_invoke, method_name):
         self.timestamp_arg_name = kw
       elif v == core.DoFn.WindowParam:
         self.window_arg_name = kw
+      elif v == core.DoFn.KeyParam:
```

Comment:
Could you also update the PerWindowInvoker (for process methods) (https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/common.py#L597)? (If possible please also update for window param that was added in the previous PR.)

Reply:
@aaltay 
I may not have a complete background but in my understanding, these parameters are mainly for the timer callback method and may not be exactly the same as the process parameter. do we still need to update `_invoke_process_per_window`? am I missing something?

Reply:
You have a valid a point. There may not always be a key that can be passed to process() method. Although sometimes there might be.

Now that we have a KeyParam, a user might right a process method using that e.g. (process(mykey=KeyParam). What would be the reasonable thing to do in this case:

1. We can do nothing and mykey will literally have the value KeyParam.
2. We can fail with an error saying that KeyParam is not a valid parameter for process method.
3. We can try to pass the key (e.g. k, v = element) and set mykey = k. And if that fails (i.e. element is not a K,V) we can fail.

I was leaning towards the third option. I think second option would also be fine and the first option will be confusing.

What do you think?

Reply:
Is there any way to know at pipeline construction time that the user is using a KV as the input type so that we don't have to defer to pipeline execution time to get an error. If no, I like Ahmets 3rd suggestion as well.

Reply:
It is possible to do this at pipeline construction time. This will be similar to how type hints detection happens. (It will be inspection at execution time vs inspection at construction time.) However we do not have the machinery in place and I suspect it will make this PR more complicated that it is intended.

We can start with a execution time error and later on improve it to be a construction time error. We can do this backward compatibly by failing in construction time only when input type hints are not in the [K, V] form, preserving the same functionality.

Reply:
I am also leaning towards option 3. If the implementation scope is bigger than I would fallback to option 2 and complete the 3rd option in a separate PR.

Reply:
The type inference machinery can tell us whether we certainly have a KV type, whether we certainly do not have a KV type, or that it doesn't know whether we have a KV type (e.g. all it knows is that it's a Python object). 

I also like option 3, and don't think it would be that hard.